doc/subsystems/jinjagen/jinjagen.rst

```diff
@@ -0,0 +1,223 @@
+configuration-specific code fragments, and more.
+
+.. image:: template_flow.png
+   :width: 500px
+   :align: center
+   :alt: Principle
+
+The code generator uses a template engine for python called jinja2. The jinja2
+templates are typically stored in files with the file extention .jinja2 which are
+rendered into other files like c-source files. The documentation for jinja2 can 
+be found here `Jinja2 <http://jinja.pocoo.org/>`_. Please refer to this site for more
+extensive information, the information in the following sections are short
+descriptions of some of the key features in jinja2. 
+
+Jinja2 files are rendered into other files using the cmake extensions functions
+
+  * zephyr_sources_codegen(..)
+  * zephyr_library_sources_codegen(..). 
+
+The generated source files are added to the Zephyr sources. During build the 
+jinja2 files are processed by jinja2 template engine and the generated source
+files are written to the CMake binary directory.
+
+A jinja2 template is written in a domain specific language, but inherited many
+consepts from python, which means that anyone familiar with python should
+learn to write jinja2 templates with ease. 
+
+Data object
+===========
+
+All templates have access to a data object called ``data``. Through this data
+object will the template have access to information like all the elements in
+the .config file, 
+
+.. code-block:: c
+
+	{{ data['config']['CONFIG_BOARD'] }}
+
+devicetree data,
+ 
+ .. code-block:: c
+
+	{% for device in data['devicetree']['devices'] %}
+	    {{ device }}
+	{% endfor %}
+
+
+and runtime information like defines that has been injected using the CODEGEN_DEFINES
+key word in the CMakeLists.txt file.
+
+.. code-block:: c
+
+	{{ data['runtime']['defines']['MY_DEFINE'] }}
+
+Filters
+=======
+
+Filters are functions that can be applied to data in jinja2 templates and are used to
+modify the data in one form or the other. The filters are applied to the data by using
+the `pipe` character 
+
+.. code-block:: c
+
+    {{ "Hello %s\\n"\|format("World!") }}
+
+
+Built-in
+--------
+
+Jinja2 has a set of built in filter that in many cases will be familiar to anyone who
+have written code in python. information about built in filter can be found here
+`Built-in Filters <http://jinja.pocoo.org/docs/2.10/templates/#list-of-builtin-filters>`_.
+
+.. code-block:: c
+
+        {{ "fOOBAR!"|capitalize() }} => Foobar!
+
+Custom
+------
+
+Filters can be added to the jinja2 engine by defining custom filters in the python
+script that call the jinja2 engine. This way can functionality easily be added to 
+the jinja2 engine to tailer it for any particular application.
+
+There is no square root filter built into jinja, this can be added by using 
+the ``sqrt()`` function in python to create a jinja2 filter
+
+Python code:
+
+.. code-block:: python
+
+        def squarerootfilter(value):
+                return sqrt(value)
+
+        environment.filters['sqrt'] = squarerootfilter
```

Comment:
Could we include here what file a user would modify to achieve this, and perhaps a disclaimer about whether it is supported to have out-of-tree filters or not?

(Unless this is documented elsewhere that I haven't seen)

Reply:
Ok, currently you add the custom filters to the main jinjagen python script in the scripts folder.

Currently out-of-tree filters are not supported, technically it is just a few simple lines of python, but conceptually it will add complexity because we will need to define some way (metadata, mechanism) to insert the out-of-tree filters and also add a version scheme for the tool it self and the filters. 

But I can add simple support for it in the prototype if you want to see how it looks/works.

Reply:
Wait until the RFC process is done